src/core/credentials/transport/xds/xds_credentials.cc

```diff
@@ -105,12 +138,20 @@ void XdsCertificateVerifier::Cancel(
 int XdsCertificateVerifier::CompareImpl(
     const grpc_tls_certificate_verifier* other) const {
   auto* o = static_cast<const XdsCertificateVerifier*>(other);
+  int compare_cert_provider;
   if (xds_certificate_provider_ == nullptr ||
       o->xds_certificate_provider_ == nullptr) {
-    return QsortCompare(xds_certificate_provider_,
-                        o->xds_certificate_provider_);
+    compare_cert_provider =
+        QsortCompare(xds_certificate_provider_, o->xds_certificate_provider_);
+  } else {
+    compare_cert_provider =
+        xds_certificate_provider_->Compare(o->xds_certificate_provider_.get());
+  }
+  if (compare_cert_provider == 0) {
+    return sni_matcher_.string_matcher().compare(
```

Comment:
This seems a little fragile, since it assumes that the two string matchers will differ only by the string, not the type of the match.  I realize that that can't happen today, but I could easily imagine someone in the future adding a new option in the ctor and forgetting to update this code.

One way to avoid this would be to wait until `Verify()` is called to construct the `StringMatcher`, so that the data member is just a string.  That's slightly sub-optimal in that it will slow things down on each individual handshake, but the effect is probably not enough to notice right now.

The alternative is to add a `Compare()` method on `StringMatcher`, but that's more work.